### PR TITLE
Added artipie base exception classes

### LINKED ISSUE #311: Base Artipie exception
Our customers work with many Artipie adapters as SDK libraries, each adapter works with exception differently. It'd be more convenient to have a base exception type to throw from SDK methods of each adapter and some additional sub-classes to provide more details (e.g. http-exception with status codes, etc).

The open questions are:
 - What is the correct repo/project to place such exception?
 - What is the name of the class (maybe `ArtipieException`)?
 - Checked or unchecked?

### PR DESCRIPTION
**Pull request by g4s8**
Fixes #311 - added `ArtipieException` base class and
`ArtipieIOException` for `asto` module, replaced
all API-related exceptions in storage implementations and
supplementary classes with these exceptions. Updated tests
to expect correct exception types on failure.

### CODE REVIEW

File: src/main/java/com/artipie/ArtipieException.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+package com.artipie;
+
+/**
+ * Base Artipie exception.
+ * <p>It should be used as a base exception for all Artipie public APIs
+ * as a contract instead of others.</p>
+ *
+ * @since 1.0
+ * @implNote ArtipieException is unchecked exception, but it's a good
+ *  practice to not it via {@code throws} tag in JavaDocs.
```
Comment (olenagerasimova) [inline src/main/java/com/artipie/ArtipieException.java:14]: @g4s8 hm, seems like there is some mistake in the sentence, I do not understand the meaning of the second part

File: src/main/java/com/artipie/asto/ArtipieIOException.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import com.artipie.ArtipieException;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+
+/**
+ * Artipie input-output exception.
+ * @since 1.0
+ * @checkstyle AbbreviationAsWordInNameCheck (10 lines)
+ */
+public class ArtipieIOException extends ArtipieException {
```
Comment (olenagerasimova) [inline src/main/java/com/artipie/asto/ArtipieIOException.java:16]: @g4s8 I'd suggest to add ctor from string and `Throwable`

**Review by olenagerasimova — CHANGES_REQUESTED** [review]
@g4s8 thanks, two comments, take a look please

**Comment by g4s8** [comment]
> @g4s8 thanks, two comments, take a look please

@olenagerasimova thanks, fixed

**Comment by olenagerasimova** [comment]
@g4s8 hm, do not see any changes

**Comment by g4s8** [comment]
> @g4s8 hm, do not see any changes

@olenagerasimova sorry, push failed, please see https://github.com/artipie/asto/pull/313/commits/69ed32d6f0f8a1bf2e312ef880adb61f834a5dfe

**Review by olenagerasimova — APPROVED** [review]
@g4s8 thanks

**Comment by g4s8** [comment]
@rultor stop

**Pull request merged**